baseapp/baseapp_test.go

```diff
@@ -2153,3 +2202,139 @@ func TestBaseApp_Init(t *testing.T) {
 		require.Equal(t, tc.expectedSnapshot.KeepRecent, tc.bapp.snapshotManager.GetKeepRecent())
 	}
 }
+
+func executeBlockWithArbitraryTxs(t *testing.T, app *BaseApp, numTransactions int, blockHeight int64) []txTest {
+	codec := codec.NewLegacyAmino()
+	registerTestCodec(codec)
+	r := rand.New(rand.NewSource(3920758213583))
+	keyCounter := 0
+	txs := make([]txTest, 0)
+
+	app.BeginBlock(abci.RequestBeginBlock{Header: tmproto.Header{Height: blockHeight}})
+	for txNum := 0; txNum < numTransactions; txNum++ {
+		tx := txTest{Msgs: []sdk.Msg{}}
+		for msgNum := 0; msgNum < 2; msgNum++ {
+			key := []byte(fmt.Sprintf("%v", keyCounter))
+			value := make([]byte, 10000)
+			_, err := r.Read(value)
+			require.NoError(t, err)
+			tx.Msgs = append(tx.Msgs, msgKeyValue{Key: key, Value: value})
+			keyCounter++
+		}
+		txBytes, err := codec.Marshal(tx)
+		require.NoError(t, err)
+		resp := app.DeliverTx(abci.RequestDeliverTx{Tx: txBytes})
+		require.True(t, resp.IsOK(), "%v", resp.String())
+		txs = append(txs, tx)
+	}
+	app.EndBlock(abci.RequestEndBlock{Height: blockHeight})
+	return txs
+}
+
+func executeBlock(t *testing.T, app *BaseApp, txs []txTest, blockHeight int64) {
+	codec := codec.NewLegacyAmino()
+	registerTestCodec(codec)
+	numTransactions := len(txs)
+
+	app.BeginBlock(abci.RequestBeginBlock{Header: tmproto.Header{Height: blockHeight}})
+	for txNum := 0; txNum < numTransactions; txNum++ {
+		tx := txs[txNum]
+
+		txBytes, err := codec.Marshal(tx)
+		require.NoError(t, err)
+		resp := app.DeliverTx(abci.RequestDeliverTx{Tx: txBytes})
+		require.True(t, resp.IsOK(), "%v", resp.String())
+	}
+	app.EndBlock(abci.RequestEndBlock{Height: blockHeight})
+}
+
+func checkSubstoreSMTsEqual(appB1 *BaseApp, appB2 *BaseApp, storeKeyName string) bool {
+	cmsB1 := appB1.cms.(*multi.Store)
+	cmsB2 := appB2.cms.(*multi.Store)
+	smtB1 := cmsB1.GetSubstoreSMT(storeKeyName)
+	smtB2 := cmsB2.GetSubstoreSMT(storeKeyName)
+
+	return string(smtB1.Root()) == string(smtB2.Root())
+}
+
+func TestGenerateAndLoadFraudProof(t *testing.T) {
+	/*
+		Tests switch between a baseapp and fraudproof and covers parts of the fraudproof cycle. Steps:
+		1. Initialize a baseapp, B1, with some state, S0
+		2. Make some state transition to state S1 by doing some transactions
+		3. Export that state S1 into a fraudProof data structure (minimal snapshot)
+		4. Load a fresh baseapp, B2, with the contents of fraud proof data structure from (3) so it can begin from state S1.
+		5. Now, pick some set of transactions, txs1, to make some more state transitions.
+		6. Execute txs1 on both B1 and B2 so they go through the same state transitions
+		7. If the state of both B1 and B2 has to converge at a state S2, then we the test passes.
+
+		Note that the appHash from B1 and B2 will not be the same because the subset of storeKeys in both apps is different
+		so we compare subStores instead
```

Comment:
Q: why are the store keys different? Shouldn't they be the same for both apps?

Reply:
the second app B2 only has a subset of the storeKeys that B1 has. Specifically, B1 has `capKey1` and `capKey2` whereas B2 only has `capKey2`.

Reply:
Why is that?

Reply:
I assume that a fraudulent block doesn't necessarily touch all the `storeKeys` of the app store which are modules in cosmos-sdk. Under that assumption, a fraudProof contains only `storeKeys` that are needed by a fraudulent block. So in this case, the `capKey2` is the only `storeKey` present in this list of `storeKeys`.

Is this assumption incorrect?

Reply:
Also, I updated this check to compare the appHash instead but with the same assumption in a PR that builds on top of this PR: https://github.com/celestiaorg/cosmos-sdk/pull/254